[Inductor] short-term fix for needs_fixed_stride_order silent incorrectness

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133639
* __->__ #133452

This is a low-risk short-term fix for
https://github.com/pytorch/pytorch/issues/128084, for the purposes of
2.4.1. The actual fix for that issue is more risky and we'll target 2.5.

needs_fixed_stride_order is silently incorrect with args that are
mutable because it creates clones of those args, writes into them, and
doesn't update the original args.

This PR makes it so that needs_fixed_stride_order doesn't apply to
inputs that are being mutated.

This PR doesn't completely fix the problem, but it makes it less
incorrect: most of the time the input already has the correct strides
but inductor fails to recognize it, and in those cases writing directly
to the input is fine.

Test Plan:
- new test

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

zou3519 committed:
[Inductor] short-term fix for needs_fixed_stride_order silent incorrectness

This is a short-term fix for
https://github.com/pytorch/pytorch/issues/128084, for the purposes of a
2.4.1.
needs_fixed_stride_order is silently incorrect with args that are
mutable because it creates clones of those args, writes into them, and
doesn't update the original args.

This PR doesn't completely fix the problem, but it makes it less
incorrect: most of the time the input already has the correct strides
but inductor fails to recognize it, and in those cases writing directly
to the input is fine.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "[Inductor] short-term fix for needs_fixed_stride_order silent incorrectness"


This is a low-risk short-term fix for
https://github.com/pytorch/pytorch/issues/128084, for the purposes of
2.4.1. The actual fix for that issue is more risky and we'll target 2.5.

needs_fixed_stride_order is silently incorrect with args that are
mutable because it creates clones of those args, writes into them, and
doesn't update the original args.

This PR makes it so that needs_fixed_stride_order doesn't apply to
inputs that are being mutated.

This PR doesn't completely fix the problem, but it makes it less
incorrect: most of the time the input already has the correct strides
but inductor fails to recognize it, and in those cases writing directly
to the input is fine.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "[Inductor] short-term fix for needs_fixed_stride_order silent incorrectness"


This is a low-risk short-term fix for
https://github.com/pytorch/pytorch/issues/128084, for the purposes of
2.4.1. The actual fix for that issue is more risky and we'll target 2.5.

needs_fixed_stride_order is silently incorrect with args that are
mutable because it creates clones of those args, writes into them, and
doesn't update the original args.

This PR makes it so that needs_fixed_stride_order doesn't apply to
inputs that are being mutated.

This PR doesn't completely fix the problem, but it makes it less
incorrect: most of the time the input already has the correct strides
but inductor fails to recognize it, and in those cases writing directly
to the input is fine.

Test Plan:
- new test

[ghstack-poisoned]

eellison commented on torch/_inductor/lowering.py:
```diff
@@ -2067,13 +2067,35 @@ def require_channels_last(_, *args, **kwargs):
+        new_args = []
+        new_kwargs = {}
+        schema_args, schema_kwargs = torch._library.utils.schema_args_kwargs(schema)
+        for arg, fx_arg, schema_arg in zip(args, fx_node.args, schema_args):
```
There is no contract that the fx node at this point is using `fx_node.args` instead of `fx_node.kwargs`. What I would expect to see here is `torch.fx.operator_schemas.normalize_function` and then iterate over the schema by argname

zou3519 replied:
Sounds good

zou3519 replied:
By the way... the code as previously written assumes that fx_node.{args, kwargs} are zip-able with {args, kwargs}. Are you saying that's incorrect too?

zou3519 replied:
I updated the code to handle the case where the schema_args/schema_kwargs can be different from args/kwargs. The way constrain_to_fx_strides is called (and from how the code was written previously) it looks like fx_node.{args, kwargs} being zip-able with {args, kwargs} is a reasonable assumption (but if we think that's a bug I can fix it in a follow-up -- I want this PR to go in without changing too much since it needs to be cherry-picked)

eellison commented on torch/_inductor/lowering.py:
```diff
@@ -2067,13 +2067,35 @@ def require_channels_last(_, *args, **kwargs):
             return ir.ExternKernel.require_stride_order(arg, stride_order)
         return arg
 
+    if ignore_mutated_args_FIXME:
```
add comment about why we're doing this ?

eellison reviewed (commented):
one comment

zou3519 committed:
Update on "[Inductor] short-term fix for needs_fixed_stride_order silent incorrectness"


This is a low-risk short-term fix for
https://github.com/pytorch/pytorch/issues/128084, for the purposes of
2.4.1. The actual fix for that issue is more risky and we'll target 2.5.

needs_fixed_stride_order is silently incorrect with args that are
mutable because it creates clones of those args, writes into them, and
doesn't update the original args.

This PR makes it so that needs_fixed_stride_order doesn't apply to
inputs that are being mutated.

This PR doesn't completely fix the problem, but it makes it less
incorrect: most of the time the input already has the correct strides
but inductor fails to recognize it, and in those cases writing directly
to the input is fine.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "[Inductor] short-term fix for needs_fixed_stride_order silent incorrectness"


This is a low-risk short-term fix for
https://github.com/pytorch/pytorch/issues/128084, for the purposes of
2.4.1. The actual fix for that issue is more risky and we'll target 2.5.

needs_fixed_stride_order is silently incorrect with args that are
mutable because it creates clones of those args, writes into them, and
doesn't update the original args.

This PR makes it so that needs_fixed_stride_order doesn't apply to
inputs that are being mutated.

This PR doesn't completely fix the problem, but it makes it less
incorrect: most of the time the input already has the correct strides
but inductor fails to recognize it, and in those cases writing directly
to the input is fine.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "[Inductor] short-term fix for needs_fixed_stride_order silent incorrectness"


This is a low-risk short-term fix for
https://github.com/pytorch/pytorch/issues/128084, for the purposes of
2.4.1. The actual fix for that issue is more risky and we'll target 2.5.

needs_fixed_stride_order is silently incorrect with args that are
mutable because it creates clones of those args, writes into them, and
doesn't update the original args.

This PR makes it so that needs_fixed_stride_order doesn't apply to
inputs that are being mutated.

This PR doesn't completely fix the problem, but it makes it less
incorrect: most of the time the input already has the correct strides
but inductor fails to recognize it, and in those cases writing directly
to the input is fine.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "[Inductor] short-term fix for needs_fixed_stride_order silent incorrectness"


This is a low-risk short-term fix for
https://github.com/pytorch/pytorch/issues/128084, for the purposes of
2.4.1. The actual fix for that issue is more risky and we'll target 2.5.

needs_fixed_stride_order is silently incorrect with args that are
mutable because it creates clones of those args, writes into them, and
doesn't update the original args.

This PR makes it so that needs_fixed_stride_order doesn't apply to
inputs that are being mutated.

This PR doesn't completely fix the problem, but it makes it less
incorrect: most of the time the input already has the correct strides
but inductor fails to recognize it, and in those cases writing directly
to the input is fine.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "[Inductor] short-term fix for needs_fixed_stride_order silent incorrectness"


This is a low-risk short-term fix for
https://github.com/pytorch/pytorch/issues/128084, for the purposes of
2.4.1. The actual fix for that issue is more risky and we'll target 2.5.

needs_fixed_stride_order is silently incorrect with args that are
mutable because it creates clones of those args, writes into them, and
doesn't update the original args.

This PR makes it so that needs_fixed_stride_order doesn't apply to
inputs that are being mutated.

This PR doesn't completely fix the problem, but it makes it less
incorrect: most of the time the input already has the correct strides
but inductor fails to recognize it, and in those cases writing directly
to the input is fine.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "[Inductor] short-term fix for needs_fixed_stride_order silent incorrectness"


This is a low-risk short-term fix for
https://github.com/pytorch/pytorch/issues/128084, for the purposes of
2.4.1. The actual fix for that issue is more risky and we'll target 2.5.

needs_fixed_stride_order is silently incorrect with args that are
mutable because it creates clones of those args, writes into them, and
doesn't update the original args.

This PR makes it so that needs_fixed_stride_order doesn't apply to
inputs that are being mutated.

This PR doesn't completely fix the problem, but it makes it less
incorrect: most of the time the input already has the correct strides
but inductor fails to recognize it, and in those cases writing directly
to the input is fine.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge -f "unrelated failure"
